### PR TITLE
feat(cosmos): Bank Precompile

### PR DESCRIPTION
**Pull request by BrickBera**

### CODE REVIEW

File: contracts/src/cosmos/precompile/bank.sol [hunk 1]
```diff
@@ -63,4 +65,120 @@ interface IBankModule {
...
+        DenomUnit[] denomUnits;
+        string base;
+        string display;
+        string name;
+        string symbol;
+    }
+
+    struct DenomsMetadata {
+        DenomMetadata[] metadatas;
+    }
```
Comment (calbera) [inline contracts/src/cosmos/precompile/bank.sol:168]: replace `DenomsMetadata` with just `[]DenomMetadata`

File: contracts/src/cosmos/precompile/bank.sol [hunk 2]
```diff
@@ -63,4 +65,120 @@ interface IBankModule {
...
+
+    struct Input {
+        address addr;
+        Coin[] coins;
+    }
+    
+    struct Output {
+        address addr;
+        Coin[] coins;
+    }
```
Comment (calbera) [inline contracts/src/cosmos/precompile/bank.sol:183]: rename Input and Output to just 1 struct, like `Balances` or UserBalances or smth

File: contracts/src/cosmos/precompile/bank.sol [hunk 3]
```diff
@@ -63,4 +65,93 @@ interface IBankModule {
...
+
+    /**
+     * @dev Send coins from one address to another.
+     */
+    function send(address fromAddress, address toAddress, Coin[] calldata amount) external payable returns (bool);
+
+    // /**
+    //  * @dev Send coins from one sender and to a series of different address. 
+    //  * If any of the receiving addresses do not correspond to an existing account, 
+    //  * a new account is created.
```
Comment (itsdevbear) [inline contracts/src/cosmos/precompile/bank.sol:121]: nit: choose a comment style

File: contracts/src/cosmos/precompile/bank.sol [hunk 4]
```diff
@@ -63,4 +65,93 @@ interface IBankModule {
...
+    struct DenomMetadata {
+        string description;
+        DenomUnit[] denomUnits;
+        string base;
+        string display;
+        string name;
+        string symbol;
+    }
+
+    // struct Balance {
```
Comment (itsdevbear) [inline contracts/src/cosmos/precompile/bank.sol:153]: commented out

File: contracts/src/cosmos/precompile/bank.sol [hunk 5]
```diff
@@ -63,4 +65,93 @@ interface IBankModule {
...
+    /**
+     * @dev Represents a cosmos coin.
+     * Note: this struct is generated as go struct that is then used in the precompile.
+     */
+    struct Coin {
+        uint256 amount;
+        string denom;
+    }
+
+    struct DenomUnit {
```
Comment (itsdevbear) [inline contracts/src/cosmos/precompile/bank.sol:138]: nit: natspec

File: cosmos/precompile/bank/bank.go [hunk 6]
```diff
@@ -21,26 +21,395 @@
...
+			Execute: c.GetSendEnabled,
+		},
+		{
+			AbiSig:  "send(address,address,(uint256,string)[])",
+			Execute: c.Send,
+		},
+	}
+}
+
+// grpc_query functions
```
Comment (itsdevbear) [inline cosmos/precompile/bank/bank.go:101]: nit: remove grpc_query functions comment

File: cosmos/precompile/bank/bank.go [hunk 7]
```diff
@@ -21,26 +21,395 @@
...
+	_ common.Address,
+	_ *big.Int,
+	readonly bool,
+	args ...any,
+) ([]any, error) {
+	addr, ok := utils.GetAs[common.Address](args[0])
+	if !ok {
+		return nil, precompile.ErrInvalidHexAddress
+	}
+	denom, ok := utils.GetAs[string](args[1])
```
Comment (itsdevbear) [inline cosmos/precompile/bank/bank.go:115]: this is going to fail linter I think

File: cosmos/precompile/bank/bank.go [hunk 8]
```diff
@@ -21,26 +21,395 @@
...
+	_ *big.Int,
+	readonly bool,
+	args ...any,
+) ([]any, error) {
+	addr, ok := utils.GetAs[common.Address](args[0])
+	if !ok {
+		return nil, precompile.ErrInvalidHexAddress
+	}
+
+	// todo: add pagination here
```
Comment (itsdevbear) [inline cosmos/precompile/bank/bank.go:146]: make a github issue and link it in the code, make it a "good open source task:"

File: cosmos/precompile/bank/bank.go [hunk 9]
```diff
@@ -21,26 +21,395 @@
...
+// GetTotalSupply implements `getTotalSupply()` method.
+func (c *Contract) GetTotalSupply(
+	ctx context.Context,
+	_ ethprecompile.EVM,
+	_ common.Address,
+	_ *big.Int,
+	readonly bool,
+	args ...any,
+) ([]any, error) {
+	// todo: add pagination here
```
Comment (itsdevbear) [inline cosmos/precompile/bank/bank.go:245]: +1 on the issue

File: cosmos/precompile/bank/bank.go [hunk 10]
```diff
@@ -21,26 +21,395 @@
...
+
+	_, err = c.msgServer.Send(ctx, &banktypes.MsgSend{
+		FromAddress: cosmlib.AddressToAccAddress(fromAddr).String(),
+		ToAddress:   cosmlib.AddressToAccAddress(toAddr).String(),
+		Amount:      coins,
+	})
+	return []any{err == nil}, err
+}
+
+// note: not doing it now, it causes too much trouble
```
Comment (itsdevbear) [inline cosmos/precompile/bank/bank.go:356]: github issue, remove the commented out code from the repo, and paste the commented code out in the github issue, reference filename and stuff

File: cosmos/precompile/bank/helpers.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+			Amount: coin.Amount.BigInt(),
+			Denom:  coin.Denom,
+		}
+	}
+	return evmCoins
+}
+
+// extractCoinsFromInput converts coins from input (of type any) into sdk.Coins.
+func extractCoinsFromInput(coins any) (sdk.Coins, error) {
+	amounts, ok := utils.GetAs[[]struct {
```
Comment (itsdevbear) [inline cosmos/precompile/bank/helpers.go:47]: maybe make a note on why this is not working? add link to the github issue you are going to make in compiler repo?

File: cosmos/precompile/bank/helpers.go [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+		Amount *big.Int `json:"amount"`
+		Denom  string   `json:"denom"`
+	}](coins)
+	if !ok {
+		return nil, precompile.ErrInvalidCoin
+	}
+
+	sdkCoins := sdk.NewCoins()
+	for _, evmCoin := range amounts {
```
Comment (itsdevbear) [inline cosmos/precompile/bank/helpers.go:56]: might be a way to do a unsafe cast here.

File: cosmos/testing/integration/precompile/bank/bank_test.go [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+
+import (
+	"fmt"
+	"math/big"
+	"os"
+	"testing"
+
+	bindings "pkg.berachain.dev/polaris/contracts/bindings/cosmos/precompile"
+
+	"pkg.berachain.dev/polaris/contracts/bindings/cosmos/precompile"
```
Comment (itsdevbear) [inline cosmos/testing/integration/precompile/bank/bank_test.go:31]: lint

File: cosmos/testing/integration/precompile/bank/bank_test.go [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+var _ = Describe("Bank", func() {
+	denom := "abera"
+	denom2 := "atoken"
+
+	BeforeEach(func() {
+
+	})
+
+	It("should call functions on the precompile directly", func() {
+
```
Comment (itsdevbear) [inline cosmos/testing/integration/precompile/bank/bank_test.go:75]: lint, new line

File: cosmos/testing/integration/precompile/bank/bank_test.go [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+		// TestAddress3 initially has 100 abera
+		balance, err := bankPrecompile.GetBalance(nil, network.TestAddress3, denom)
+		Expect(err).ShouldNot(HaveOccurred())
+		Expect(balance).To(Equal(big.NewInt(100)))
+
+		// Send 1000 bera from TestAddress to TestAddress3
+		tx, err := bankPrecompile.Send(btf.GenerateTransactOpts(""), network.TestAddress, network.TestAddress3, coinsToBeSent)
+		Expect(err).ShouldNot(HaveOccurred())
+		fmt.Printf("tx is: %v", *tx)
```
Comment (itsdevbear) [inline cosmos/testing/integration/precompile/bank/bank_test.go:113]: lint

File: cosmos/testing/integration/precompile/bank/bank_test.go [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+		Expect(spendableBalances).To(Equal(expectedAllBalance))
+
+		atokenSupply, err := bankPrecompile.GetSupply(nil, denom2)
+		Expect(err).ShouldNot(HaveOccurred())
+		Expect(atokenSupply).To(Equal(big.NewInt(100)))
+
+		totalSupply, err := bankPrecompile.GetAllSupply(nil)
+		Expect(err).ShouldNot(HaveOccurred())
+		fmt.Printf("\ntotalSupply is: %v\n", totalSupply)
+
```
Comment (itsdevbear) [inline cosmos/testing/integration/precompile/bank/bank_test.go:144]: lint

File: cosmos/testing/integration/precompile/bank/bank_test.go [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+		denomMetadata, err := bankPrecompile.GetDenomMetadata(nil, denom)
+		Expect(err).ShouldNot(HaveOccurred())
+		fmt.Println("denomMetadata is: ", denomMetadata)
```
Comment (itsdevbear) [inline cosmos/testing/integration/precompile/bank/bank_test.go:147]: lint

File: cosmos/testing/integration/precompile/staking/staking_test.go [hunk 18]
```diff
@@ -37,24 +37,24 @@ import (
...
-func TestCosmosPrecompiles(t *testing.T) {
+func TestStakingPrecompile(t *testing.T) {
 	RegisterFailHandler(Fail)
-	RunSpecs(t, "cosmos/testing/precompile:integration")
+	RunSpecs(t, "cosmos/testing/integration/precompile/staking")
 }
 
 var (
-	tf                *integration.TestFixture
+	stf               *integration.TestFixture
```
Comment (itsdevbear) [inline cosmos/testing/integration/precompile/staking/staking_test.go:46]: why stf?

File: cosmos/testing/network/network.go [hunk 19]
```diff
@@ -63,12 +63,20 @@ const (
...
-	ECDSATestKey, _ = TestKey.ToECDSA()
-	AddressFromKey  = TestKey.PubKey().Address()
-	Signer          = coretypes.LatestSignerForChainID(params.DefaultChainConfig.ChainID)
-	TestAddress     = crypto.PubkeyToAddress(ECDSATestKey.PublicKey)
+	DummyContract    = common.HexToAddress("0x9fd0aA3B78277a1E717de9D3de434D4b812e5499")
+	TestKey, _       = ethsecp256k1.GenPrivKey()
+	TestKey2, _      = ethsecp256k1.GenPrivKey()
+	TestKey3, _      = ethsecp256k1.GenPrivKey()
+	ECDSATestKey, _  = TestKey.ToECDSA()
+	ECDSATestKey2, _ = TestKey2.ToECDSA()
```
Comment (itsdevbear) [inline cosmos/testing/network/network.go:71]: refactor this into the fixture and make an array of addresses, as I mentioned before

File: cosmos/testing/network/network.go [hunk 20]
```diff
@@ -168,3 +207,32 @@ func BuildGenesisState() map[string]json.RawMessage {
...
+			Name:        "Token",
+			Symbol:      "TOKEN",
+			Description: "The native staking token of the Token Hub.",
+			DenomUnits: []*banktypes.DenomUnit{
+				{Denom: "1token", Exponent: uint32(5), Aliases: []string{"decitoken"}},
+				{Denom: "2token", Exponent: uint32(4), Aliases: []string{"centitoken"}},
+				{Denom: "3token", Exponent: uint32(7), Aliases: []string{"dekatoken"}},
+			},
+			Base:    "utoken",
+			Display: "token",
```
Comment (itsdevbear) [inline cosmos/testing/network/network.go:235]: bunch of linter errors inroduced into this file, i'd handle in the fixture

**Review by itsdevbear — CHANGES_REQUESTED** [review]
few things

File: cosmos/testing/integration/precompile/staking/staking_test.go (hunk 18 above)
Comment (calbera) [reply cosmos/testing/integration/precompile/staking/staking_test.go:46]: yea now that its in different packages, each one can have `tf`

File: cosmos/testing/integration/precompile/bank/bank_test.go [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+			{
+				Denom:  denom2,
+				Amount: big.NewInt(100),
+			},
+		}
+		evmDenomMetadata := precompile.IBankModuleDenomMetadata{
+			Name:        "Berachain bera",
+			Symbol:      "BERA",
+			Description: "The Bera.",
+			DenomUnits: []precompile.IBankModuleDenomUnit{
```
[the remark this answers is not in the document]
Comment (BrickBera) [reply cosmos/testing/integration/precompile/bank/bank_test.go:96]: the bank module type has `DenomUnits []*DenomUnit`, it cannot be casted

**Comment by BrickBera** [comment]
created Github issues: 
https://github.com/berachain/polaris/issues/547
https://github.com/berachain/polaris/issues/548

File: contracts/src/cosmos/precompile/examples/Fundraiser.sol [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+        bank.send(msg.sender, address(this), coins);
+    }
+    
+    function GetRaisedAmounts() public view returns (IBankModule.Coin[] memory){
+        return bank.getAllBalances(address(this));
+    }
+
+    receive() external payable {
+        // this built-in function doesn't require any calldata,
+        // it will get called if the data field is empty and
```
Comment (itsdevbear) [inline contracts/src/cosmos/precompile/examples/Fundraiser.sol:71]: can probably remove this comment

File: cosmos/testing/integration/fixture.go [hunk 23]
```diff
@@ -38,15 +39,20 @@ import (
...
+
 // TestFixture is a testing fixture that can be used to test the
 // Ethereum JSON-RPC API.
 type TestFixture struct {
 	t           network.TestingT
 	Network     *network.Network
 	EthClient   *ethclient.Client
 	EthWsClient *ethclient.Client
-	HTTPAddr    string
+  HTTPAddr    string
```
Comment (itsdevbear) [inline cosmos/testing/integration/fixture.go:53]: fix space (my bad :P)

File: cosmos/testing/integration/fixture.go [hunk 24]
```diff
@@ -93,10 +104,10 @@ func (tf *TestFixture) GenerateTransactOpts(name string) *bind.TransactOpts {
 	if err != nil {
 		tf.t.Fatal(err)
 	}
-	// nonce, err := client.PendingNonceAt(context.Background(), network.TestAddress)
+	// nonce, err := client.PendingNonceAt(context.Background(), tf.Address("MainAcc"))
```
Comment (itsdevbear) [inline cosmos/testing/integration/fixture.go:107]: MainAcc kinda weird,   

I'd do like `alice` `bob` ..... the standard cs ones

**Review by itsdevbear — APPROVED** [review]
looks really good, just a couple minor things

utACK

File: contracts/src/cosmos/precompile/examples/Fundraiser.sol [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+    // State
+    IBankModule public immutable bank = IBankModule(0x4381dC2aB14285160c808659aEe005D51255adD7);
+
+    event Success(bool indexed success);
+    event Data(bytes data);
+
+    // Errors
+    error ZeroAddress();
+    error ZeroAmount();
+    error InvalidValue();
```
Comment (calbera) [inline contracts/src/cosmos/precompile/examples/Fundraiser.sol:50]: can remove these, unused

File: cosmos/precompile/bank/bank.go [hunk 26]
```diff
@@ -21,26 +21,332 @@
...
 type Contract struct {
 	precompile.BaseContract
+
+	msgServer banktypes.MsgServer
+	querier   banktypes.QueryServer
 }
 
 // NewPrecompileContract returns a new instance of the bank precompile contract.
-func NewPrecompileContract() ethprecompile.StatefulImpl {
+func NewPrecompileContract(bk bankkeeper.Keeper) ethprecompile.StatefulImpl {
```
Comment (calbera) [inline cosmos/precompile/bank/bank.go:48]: you want to pass in the interfaces in the function

File: contracts/src/cosmos/precompile/examples/Fundraiser.sol [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+/**
+ * @dev LiquidStaking is a contract that allows users to delegate their Base Denom to a validator
+ * and receive a liquid staking token in return. The liquid staking token can be redeemed for Base
+ * Denom at any time.
+ * Note: This is an example of how to delegate Base Denom to a validator.
+ * Doing it this way is unsafe since the user can delegate more straight through precomile.
+ * And withdraw via the precompile.
+ */
+contract Fundraiser {
+    address public owner;
```
Comment (calbera) [inline contracts/src/cosmos/precompile/examples/Fundraiser.sol:39]: can remove this field and use the `Owner` contract

Comment (BrickBera) [reply contracts/src/cosmos/precompile/examples/Fundraiser.sol:39]: what is "the Owner contract"?

**Review by calbera — APPROVED** [review]

**Pull request merged**